docs/zkapps/o1js/bitwise-operations.mdx

```diff
@@ -0,0 +1,296 @@
+keywords:
+  - gadgets
+  - bitwise operations
+  - o1js
+---
+
+:::info
+
+Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.
+
+:::
+
+# Bitwise Operations
+
+In o1js, bitwise operations are implemented as [gadgets](/zkapps/o1js/gadgets).
+
+- [and()](#and)
+- [not()](#not)
+- [xor()](#xor)
+- [leftShift()](#leftshift)
+- [rightShift()](#rightshift)
+- [rotate()](#rotate)
+- [rangeCheck64()](#rangecheck64)
+- [multiRangeCheck()](#multirangecheck)
+- [compactMultiRangeCheck()](#compactmultirangecheck)
+
+See the type declaration for [Gadgets](/zkapps/o1js-reference/modules#gadgets) in the o1js Reference documentation.
+
+## and()
+
+```ts
+and(a: Field, b: Field, length: number) => Field
+```
+
+The bitwise `and()` gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 
+
+The `and()` gadget:
+
+- Compares two bits and returns 1 only if both bits are 1. Returns 0 if either bit is not 1.
+- Constrains both input values to fit into `paddedLength` bits. 
+- Both Field elements must fit into `2^paddedLength - 1`.
+- Guarantees that the output is has at most `paddedLength` bits.
+
+For details about the implementation, see [AND](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#and) in the Mina book.
+
+The `length` parameter:
+
+- Defines how many bits to compare. 
+- Rounds `length` to the nearest multiple of 16, `paddedLength = ceil(length / 16) * 16`.
+- A larger `length` parameter adds more constraints.
+
+With `length = 2` (`paddedLength = 16`), the `and()` operation fails for any input that is larger than `2**16`.
+
+Example:
+
+```ts
+let a = Field(3);    // ... 000011
+let b = Field(5);    // ... 000101
+let c = Gadgets.and(a, b, 2);    // ... 000001
+c.assertEquals(1);
+```
+
+## not()
+
+```ts
+not(a: Field, length: number, checked: boolean) => Field
+```
+
+The bitwise `not()` gadget on Field elements is similar to the [Bitwise NOT (~)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_NOT) operator in JavaScript. It is a provable method to support bitwise shifting for native Field elements.
+
+A bitwise NOT operation returns 1 in each bit position if the corresponding bit of the operand is 0 and returns 0 if the corresponding bit of the operand is 1.
+
+The operation:
+
+- Fails if the length or the input value is larger than 254 bits.
+- Operates only over the number of bits specified by the `length` parameter.
+    - The `length` parameter defines how many bits to NOT.
+    - A larger `length` parameter adds more constraints.
+- Is implemented in two different ways:
+    - If the `checked` parameter is set to `false`, it is implemented as a subtraction of the input from the all one bitmask. If the `checked` parameter is not provided, the default implementation uses `false`.
```

Comment:
"from the all one bitmask" seems like an error but I don't know what it should say